### PR TITLE
Fix #3579: Add Volcano extenstion

### LINKED ISSUE #3579: Add support for Volcano extension
Volcano is a Kubernetes native system for high-performance workloads, which has been accepted by Cloud Native Computing Foundation (CNCF) as its first and only official container batch scheduling project.

There were some initial discussion on https://github.com/fabric8io/kubernetes-client/issues/3043 .

This issue would be a start and help to track the process on Volcano extension.

[1] https://volcano.sh/en/
[2] https://github.com/volcano-sh/volcano/issues/1825

### PR DESCRIPTION
**Pull request by Yikun**
## Description
Volcano is a Kubernetes native system for high-performance workloads, which has been accepted by Cloud Native Computing Foundation (CNCF) as its first and only official container batch scheduling project.

This patch adds the volcano extension support, the user can use below code to list API using v1beta1 API:
```java
NamespacedVolcanoClient volcanoClient = new DefaultVolcanoClient()
// Current, will call v1beta1 by default
volcanoClient.podGroups().inNamespace("default").list()
// You can also sepecify the certion version
volcanoClient.v1beta1().podGroups().inNamespace("default").list()

```

Changes:
- Add `generate.go` to generate `volcano-schema.json` from [volcano-sh/apis](https://github.com/volcano-sh/apis).
- Generate Java Volcano model according `volcano-schema.json`, the v1beta1 include `PodGroup`, `Queue` related scheduling API.
- Add volcano client (default and v1beta1).
- Mock, test and examples.

Fix https://github.com/fabric8io/kubernetes-client/issues/3579

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [x] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [x] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [x] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by Yikun** [comment]
Looks like CI failed is unrelated, so I think the PR it's ready for review now.

**Comment by rohanKanojia** [comment]
@matthyx: You have contributed a few extensions till now so I think you're also familiar with the extensions module. Would appreciate it if you could help us out in reviewing this PR whenever you get time :-)

**Comment by dipanjank** [comment]
Thanks for this! One question - I don't see the Java equivalent for Job or Tasks. Are the users of this component supposed to use PodGroup instead?

**Comment by Yikun** [comment]
> Thanks for this! One question - I don't see the Java equivalent for Job or Tasks. Are the users of this extension supposed to use PodGroup instead?

Good question, The job/taks related resource is belong to `batch/v1alpha1`. As mentioned in commit message, this PR is only cover the `scheduling v1beta1` API as initial proposal, I'm also going to propose the separate PR to cover `batch/v1alpha1`, `bus/v1alpha1`, `nodeinfo/v1alpha1` API, by this way, it is more clearer and more convenient to review.

**Comment by dipanjank** [comment]
> > Thanks for this! One question - I don't see the Java equivalent for Job or Tasks. Are the users of this extension supposed to use PodGroup instead?
> 
> Good question, The job/taks related resource is belong to `batch/v1alpha1`. As mentioned in commit message, this PR is only cover the `scheduling v1beta1` API as initial proposal, I'm also going to propose the separate PR to cover `batch/v1alpha1`, `bus/v1alpha1`, `nodeinfo/v1alpha1` API, by this way, it is more clearer and more convenient to review.

Ah okay, agreed! Nice work :)

File: extensions/volcano/generator-v1beta1/cmd/generate/generate.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"reflect"
+	volcanov1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+)
+
+func main() {
+
+	// the CRD List types for which the model should be generated
+	// no other types need to be defined as they are auto discovered
+	crdLists := map[reflect.Type]schemagen.CrdScope{
+		reflect.TypeOf(volcanov1beta1.PodGroupList{}):   schemagen.Namespaced,
```
Comment (matthyx) [inline extensions/volcano/generator-v1beta1/cmd/generate/generate.go:32]: gofmt is off here... please run `gofmt -w extensions/volcano/generator-v1beta1/cmd/generate/generate.go`

File: extensions/volcano/model-v1beta1/pom.xml [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+              </artifacts>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-gpg-plugin</artifactId>
+        <configuration>
+          <passphrase>4uds0n</passphrase>
```
Comment (matthyx) [inline extensions/volcano/model-v1beta1/pom.xml:142]: This is no longer needed IMHO.

**Review by matthyx — COMMENTED** [review]
Looks good, only nitpicking... congrats!

File: extensions/volcano/generator-v1beta1/cmd/generate/generate.go (hunk 1 above)
Comment (Yikun) [reply extensions/volcano/generator-v1beta1/cmd/generate/generate.go]: Thanks! Done.

File: extensions/volcano/model-v1beta1/pom.xml (hunk 2 above)
Comment (Yikun) [reply extensions/volcano/model-v1beta1/pom.xml:142]: I'm not sure about this, let's see others suggestions.

If it should be cleanup, looks like passphrase hard code should be cleanup both in [root pom](https://github.com/fabric8io/kubernetes-client/blob/4f2d12cc5ecc72b4dfc4b02c95a3715d8dd438f5/pom.xml#L870) and [all extensions](https://github.com/fabric8io/kubernetes-client/search?q=4uds0n)? If it is, I'd like to submit a separated PR to cleanup these.

**Comment by Yikun** [comment]
I noticed 5.10 had been released, bump dep version to v5.11

File: extensions/volcano/model-v1beta1/pom.xml (hunk 2 above)
Comment (Yikun) [reply extensions/volcano/model-v1beta1/pom.xml:142]: Maybe cc @rohanKanojia @manusa do you have any input in here?

**Comment by Yikun** [comment]
Any more suggestion on this PR? Thanks!

**Comment by matthyx** [comment]
Looks good to me, thanks!

**Comment by Yikun** [comment]
@matthyx Thanks for your help!

**Comment by Yikun** [comment]
@rohanKanojia Would you mind take a look again on this? Thanks!

File: extensions/volcano/tests/pom.xml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      <groupId>org.slf4j</groupId>
+      <artifactId>slf4j-simple</artifactId>
+      <version>${slf4j.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>volcano-model-v1beta1</artifactId>
+      <version>5.11-SNAPSHOT</version>
+      <scope>test</scope>
```
Comment (rohanKanojia) [inline extensions/volcano/tests/pom.xml:67]: Why is this dependency required?

File: extensions/volcano/tests/pom.xml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  <artifactId>volcano-tests</artifactId>
+  <packaging>jar</packaging>
+  <name>Fabric8 :: Volcano :: Tests</name>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>kubernetes-client</artifactId>
+    </dependency>
```
Comment (rohanKanojia) [inline extensions/volcano/tests/pom.xml:36]: Is this required?

File: extensions/volcano/client/pom.xml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      <artifactId>kubernetes-client</artifactId>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-migrationsupport</artifactId>
```
Comment (rohanKanojia) [inline extensions/volcano/client/pom.xml:75]: We had this in extensions to use JUnit4  `@Rule` for Mock Server. Now that you're using JUnit5 annotation in your tests this should not be required

File: extensions/volcano/mock/pom.xml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-migrationsupport</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Comment (rohanKanojia) [inline extensions/volcano/mock/pom.xml:52]: Not required

File: extensions/volcano/tests/pom.xml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      <artifactId>volcano-server-mock</artifactId>
+      <version>${project.version}</version>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>mockwebserver</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Comment (rohanKanojia) [inline extensions/volcano/tests/pom.xml:49]: I think you can remove this also. It's already coming as a transitive dependency from `kubernetes-server-mock`

File: extensions/volcano/tests/src/test/java/io/fabric8/volcano/test/V1Beta1VolcanoTest.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+import io.fabric8.volcano.scheduling.v1beta1.PodGroupBuilder;
+import io.fabric8.volcano.scheduling.v1beta1.QueueBuilder;
+import io.fabric8.volcano.server.mock.EnableVolcanoMockClient;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@EnableVolcanoMockClient(crud = true)
+class V1Beta1VolcanoTest {
```
Comment (rohanKanojia) [inline extensions/volcano/tests/src/test/java/io/fabric8/volcano/test/V1Beta1VolcanoTest.java:28]: CRUD mode tests are good but I would appreciate if you could add some expectations mode tests as well. They help in verifying that correct URLs are being hit and correct plural is being calculated by client.

**Comment by rohanKanojia** [comment]
Could you please update this Extensions table in the README as well: https://github.com/fabric8io/kubernetes-client/blob/master/README.md ?

**Comment by Yikun** [comment]
@rohanKanojia
1. For pom issue, I recheck all pom, and finally get below deps tree:
```
volcano-client --> volcano-model-v1beta1
volcano-examples --> volcano-client
volcano-server-mock --> volcano-client
volcano-tests --> volcano-client + volcano-server-mock
```

2. For other test besides curd
> CRUD mode tests are good but I would appreciate if you could add some expectations mode tests as well. They help in verifying that correct URLs are being hit and correct plural is being calculated by client.

I added io.fabric8.volcano.test.request.[PodGroupTest | QueueTest] to cover 

3. readme updated

Much thanks for your suggestion!

File: extensions/volcano/tests/src/test/java/io/fabric8/volcano/test/V1Beta1VolcanoTest.java (hunk 8 above)
Comment (Yikun) [reply extensions/volcano/tests/src/test/java/io/fabric8/volcano/test/V1Beta1VolcanoTest.java]: will add some https mode test

**Comment by Yikun** [comment]
- `WatchTest.testHttpErrorReconnect:193 expected: <true> but was: <false>` Java 11 maven failed is unrelated.
- SonarCloud failed due to same reason ""passphrase" detected here, make sure this is not a hard-coded credential." mentioned previously.

No more changes from me, so it's ready for review.

**Comment by william-wang** [comment]
I am volcano communtiy maintainer. It's awesome that volcano extension can be supported in kubenetes-clients. As really a lot of users has this need.

**Comment by william-wang** [comment]
I reviewed the code. It looks good to me.

**Comment by Yikun** [comment]
> I reviewed the code. It looks good to me.

Thanks for review!

**Comment by Yikun** [comment]
@rohanKanojia Any more suggestion? Thanks!

**Review by rohanKanojia — APPROVED** [review]

**Review by oscerd — APPROVED** [review]

**Review by manusa — APPROVED** [review]
LGTM, thx!

**Comment by Yikun** [comment]
@rohanKanojia @matthyx @oscerd @manusa @william-wang @dipanjank Thanks for your help and review, Thank you guys!

**Pull request merged**